### PR TITLE
Feature transformer

### LINKED ISSUE #376: Per class feature constructor
Create a class that perform the same task as the current `DDGTransformer`, but with a transformer instead of depths. The class will just receive one scikit-learn-like transformer (with `fit` and `transform`) and will train and transform in each class separatedly, returning the obtained set of features.

We should prevent the users if we can for passing a transformer that do not use target data in `fit`. Otherwise there would be repeated columns in the output. That could be done by checking the `requires_y` and the `stateless` estimator tags (see https://scikit-learn.org/stable/developers/develop.html#estimator-tags) and raising an exception or logging a warning. We should try to add the tags to all our models in the future.

### PR DESCRIPTION
**Pull request by alvaro-castillo**
Closes #376

### CODE REVIEW

**Review by vnmabus — CHANGES_REQUESTED** [review]
Apply also the same changes as in the other PR, please.

**Review by alvaro-castillo — COMMENTED** [review]
I have corrected the code

**Comment by alvaro-castillo** [comment]
All errors from mypy on _utils are on functions that were already implemented. I just added from line 736 to 748 and style errors appear in other lines. Should I bypass the warnings by commenting #noqa: ....?

File: tests/test_per_class_transformer.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+from skfda.ml.classification import KNeighborsClassifier
+from skfda.preprocessing.dim_reduction.feature_extraction import (
+    PerClassTransformer,
+)
+from skfda.preprocessing.dim_reduction.variable_selection import (
+    RecursiveMaximaHunting,
+)
+
+
+class TestPCT(unittest.TestCase):
```
Comment (vnmabus) [inline tests/test_per_class_transformer.py:18]: Don't call it PCT. It is confusing.

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+import numpy as np
+from pandas import DataFrame
+from sklearn.utils.validation import check_is_fitted as sklearn_check_is_fitted
+
+from ...._utils import TransformerMixin, _fit_feature_transformer
+from ....representation._typing import NDArrayInt
+from ....representation.basis import FDataBasis
+from ....representation.grid import FData, FDataGrid
+
+T = TypeVar("T", bound=FData)
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py:16]: Remove this.

File: skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+            warnings.warn(
+                f"Parameter ``transformer`` with type"  # noqa: WPS237
+                f" {type(self.transformer)} should use class information."
+                f" It should have the ``requires_y`` tag set to ``True`` and"
+                f" the ``stateless`` tag set to ``False``",
+            )
+
+    def fit(
+        self,
+        X: T,
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py:141]:
```suggestion
        X: Input,
```

File: skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        y: np.ndarray,
+    ) -> PerClassTransformer:
```
[the remark this answers is not in the document]
Comment (vnmabus) [reply skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py:143]:
```suggestion
    ) -> PerClassTransformer[Input, Output, Target]:
```

File: skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            for i in transformed_data:
+                if isinstance(i, (FDataGrid, FDataBasis)):
+                    raise TypeError(
+                        "There are transformed instances of FDataGrid or "
+                        "FDataBasis that can't be concatenated on a NumPy "
+                        "array.",
+                    )
+            return np.array(transformed_data)
+
+        return DataFrame(
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py:197]: You should test also the `DataFrame` case, ideally with transformers returning `FData`.

File: skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+            {'Transformed data': transformed_data},
+        )
+
+    def fit_transform(
+        self,
+        X: T,
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py:203]:
```suggestion
        X: Input,
```

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py [hunk 7]
```diff
@@ -178,10 +235,16 @@ def transform(self, X: T) -> Union[DataFrame, np.ndarray]:
         """
         sklearn_check_is_fitted(self)
 
-        transformed_data = np.empty((len(X), 0))
-        for feature_transformer in self._class_feature_transformers_:
+        for j, feature_transformer in enumerate(
+            self._class_feature_transformers_,
+        ):
             elem = feature_transformer.transform(X)
             data = np.array(elem)
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py:242]: Why? Why not simply putting every column in a list using a list comprehension and THEN converting to array/dataframe?

Comment (alvaro-castillo) [reply skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py:242]: I have just changed it to a list comprehension.
While creating a ndarray as output with np.array(transformed_data) an exception is risen: ValueError: could not broadcast input array from shape (x) into shape (y), and so I have also changed the way the ndarray is created.

File: skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+                if isinstance(data, (FDataGrid, FDataBasis)):
+                    raise TypeError(
+                        "There are transformed instances of FDataGrid or "
+                        "FDataBasis that can't be concatenated on a NumPy "
+                        "array.",
+                    )
+                elif i == 0:
+                    transformed_array = data
+                else:
+                    transformed_array = np.hstack((transformed_array, data))
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py:254]: Can't you do `np.array(transformed_data).T`? Concatenating arrays is absurd if you already have all the data.

Comment (alvaro-castillo) [reply skfda/preprocessing/dim_reduction/feature_extraction/_per_class_transformer.py:254]: The error still appears.
When executing pytest, the first example shows: "ValueError('could not broadcast input array from shape (93,24) into shape (93,)')" .
It seems that the arrays contained in transformed_data have sizes (93, 24) and (93, 18) and that is why it gives an exception when creating a ndarray.

**Review by vnmabus — APPROVED** [review]
LGTM

**Pull request merged**